include/deal.II/fe/fe_interface.h

```diff
@@ -0,0 +1,787 @@
+   * underlying FEFaceValues.
+   */
+  unsigned int
+  interface_dof_index_to_fe_dof_index(unsigned int interface_dof_index) const
+  {
+    AssertIndexRange(interface_dof_index, n_interface_dofs());
+    return (interface_dof_index >= n_dofs_fe) ?
+             (interface_dof_index - n_dofs_fe) :
+             interface_dof_index;
+  }
+
+  /**
+   * For a given interface dof index return whether it belongs to the cell (0)
+   * or the neighbor (1).
+   */
+  unsigned int
+  interface_dof_index_to_fe_index(const unsigned int interface_dof_index) const
+  {
+    AssertIndexRange(interface_dof_index, n_interface_dofs());
+    return (interface_dof_index < n_dofs_fe) ? 0 : 1;
+  }
+
+  /**
+   * Convert an FiniteElement index (0=cell, 1=neighbor) and dof index @p face_dof_index
+   * into a interface DoF index.
+   *
+   * The inverse of this operation is done with
+   * interface_dof_index_to_fe_index() and interface_dof_idx_to_fe_dof_idx().
+   */
+  unsigned int
+  interface_dof_index(const unsigned int fe_index,
+                      const unsigned int face_dof_index) const
+  {
+    Assert(fe_index <= 1, ExcMessage("fe_index should be 0 or 1"));
+    if (is_boundary_facet())
+      Assert(fe_index == 0,
+             ExcMessage("A boundary facet only has FE index 0."));
+
+    if (fe_index == 0)
+      {
+        Assert(face_dof_index < n_dofs_fe, ExcMessage("invalid face_dof_idx"));
+        return face_dof_index;
+      }
+    else if (fe_index == 1)
+      {
+        Assert(face_dof_index < n_dofs_fe_neighbor,
+               ExcMessage("invalid face_dof_idx"));
+        return face_dof_index + n_dofs_fe;
+      }
+    // return something invalid:
+    return -1;
+  }
+
+  /**
+   * Return the FEFaceValue object of the cell or the neighboring cell
+   */
+  const FEFaceValuesBase<dim, spacedim> &
+  get_fe_values(const unsigned int cell_or_neighbor) const
+  {
+    AssertIndexRange(cell_or_neighbor, 2);
+    return (cell_or_neighbor == 0) ? get_fe_values() : get_fe_values_neighbor();
+  }
+
+  const FEFaceValuesBase<dim, spacedim> &
+  get_fe_values() const
+  {
+    return *fe_face_values;
+  }
+
+  const FEFaceValuesBase<dim, spacedim> &
+  get_fe_values_neighbor() const
+  {
+    Assert(!is_boundary_facet(),
+           ExcMessage("Not possible for boundary Facet."));
+    return *fe_face_values_neighbor;
+  }
+
+  /**
+   * Return the normal in a given quadrature point.
+   */
+  Tensor<1, spacedim>
+  normal(const unsigned int q_point_index) const
+  {
+    return fe_face_values->normal_vector(q_point_index);
+  }
+
+  /**
+   * Return a view of scalar FE.
+   */
+  const FEInterfaceViews::Scalar<dim, spacedim> &
+  operator[](const FEValuesExtractors::Scalar &scalar) const
+  {
+    Assert(scalar.component < fe_face_values->get_fe().n_components(),
+           ExcMessage("Invalid FEValuesExtractors::Scalar!"));
+
+    return cached_views_scalar[scalar.component];
+  }
+
+  /**
```

Comment:
I'm not sure about this. I would remove the `scalar()` operator. What do others think?